ops/pebble.py

```diff
@@ -308,6 +317,82 @@ def __repr__(self):
                 ).format(self=self)
 
 
+class Config:
+    """Represents a Pebble config layer (or flattened configuration).
+
+    This is not documented, but captured in code here:
+    https://github.com/canonical/pebble/blob/master/internal/setup/setup.go
+    """
+
+    def __init__(self, raw: Union[str, Dict]):
+        if isinstance(raw, str):
+            d = yaml.safe_load(raw)
+        else:
+            d = raw
+        self.summary = d.get('summary', '')
+        self.description = d.get('description', '')
+        self.services = {name: ConfigService(name, service)
+                         for name, service in d.get('services', {}).items()}
+
+    def to_yaml(self) -> str:
+        """Convert this config to its YAML representation."""
+        return yaml.dump(self.to_dict(), Dumper=_DefaultDumper)
+
+    def to_dict(self) -> Dict:
+        """Convert this config to a dict representation."""
+        return {
+            'summary': self.summary,
+            'description': self.description,
+            'services': {name: service.to_dict() for name, service in self.services.items()},
+        }
+        raise NotImplementedError
+
+    def __repr__(self) -> str:
+        return 'Config({!r})'.format(self.to_dict())
+
+    __str__ = to_yaml
+
+
+class ConfigService:
+    """Represents a service description in a Pebble configuration layer."""
+
+    def __init__(self, name: str, raw: Dict):
+        self.name = name
+        self.summary = raw.get('summary', '')
+        self.description = raw.get('description', '')
+        self.default = raw.get('default', '')
+        self.override = raw.get('override', '')
+        self.command = raw.get('command', '')
+        self.after = raw.get('after', [])
+        self.before = raw.get('before', [])
+        self.requires = raw.get('requires', [])
+        self.environment = raw.get('environment') or {}
+
+    def to_yaml(self) -> str:
+        """Convert this service object to its YAML representation."""
+        return yaml.dump(self.to_dict(), Dumper=_DefaultDumper)
+
+    def to_dict(self) -> Dict:
+        """Convert this service object to its dict representation."""
+        return {
+            'name': self.name,
+            'summary': self.summary,
+            'description': self.description,
+            'default': self.default,
+            'override': self.override,
+            'command': self.command,
+            'after': self.after,
+            'before': self.before,
+            'requires': self.requires,
+            'environment': self.environment,
+        }
+
+    def __repr__(self) -> str:
+        return 'ConfigService({!r})'.format(self.to_dict())
+
+    __str__ = to_yaml
+
+
 class API:
     """Pebble API client."""
```

Comment:
Commenting here because I can't comment on code below as it was already there: the error message below isn't following the lowcase convention, and it seems a bit unusual in that it addresses the user directly with an order ("You must"). It's also improper in that it talks to the user about a variable name in the code.

It would be nicer as, for example:

"cannot communicate with pebble socket (no path provided and $PEBBLE not set)"

Reply:
Fixed. I've also taken the environment variable handling out of the class -- you explicitly pass in a `socket_path`, and the calling code (e.g., the command line test client in pebble.py) does environment variable handling if it needs it.